src/main/java/com/code_intelligence/jazzer/android/AndroidRuntime.java

```diff
@@ -0,0 +1,56 @@
+// Copyright 2021 Code Intelligence GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.code_intelligence.jazzer.android;
+
+import com.code_intelligence.jazzer.driver.Opt;
+import com.code_intelligence.jazzer.utils.Log;
+import com.github.fmeum.rules_jni.RulesJni;
+import java.io.UnsupportedEncodingException;
+
+/**
+ * Loads Android tooling library and registers native functions.
+ */
+public class AndroidRuntime {
+  private static final String doNotInitialize = "use_none";
+
+  public static void initialize(String runtimeLibs) throws UnsupportedEncodingException {
+    if (runtimeLibs == null) {
+      return;
+    }
+
+    if (Opt.isAndroid) {
+      RulesJni.loadLibrary("jazzer_android_tooling", "/com/code_intelligence/jazzer/driver");
+      //   RulesJni.loadLibrary("jazzer_android_tooling", AndroidRuntime.class);
+      if (!runtimeLibs.equals(doNotInitialize)) {
+        byte[] bytes = runtimeLibs.getBytes("UTF-8");
+        registerNatives(bytes);
```

Comment:
Why convert this to bytes and then convert back to string in the native function?

I think this should be:
private static native int registerNatives(String runtimeLibs)

and 
Java_com_code_1intelligence_jazzer_android_AndroidRuntime_registerNatives(JNIEnv *env, jclass clazz, jstring runtimeLib)

Reply:
Believe it or not, this is the original way I implemented it. TL;DR -- Java & Android have different character encodings (UTF-16 & UTF-8, respectively).

More detailed answer can be found in [this stackoverflow thread](https://stackoverflow.com/questions/20536296/jni-call-convert-jstring-to-char):

```
5. The idea of converting a JVM object reference (jstring) to a pointer to one byte storage (char*) needs a lot of refinement. You probably want to copy the characters in a JVM java.lang.String to a "native" string using a specific or OS-default encoding. The Java string has Unicode characters with a UTF-16 encoding. Android generally uses the Unicode character set with the UTF-8 encoding.
```

Note that I implemented the answerer's suggestion:

```
However, I'd probably do the call to String.getBytes on the Java side, defining the native method to take a byte array instead of a string.
```

To summarize, the code was more complicated this way. I can change it back if you still think it would be better, but hopefully this explains how I landed on this decision.